Some procs to deal with Rune position base indexing.

oderwat opened the pull request:
It can't be perfect but at least one can index on rune position
efficiently.

Araq commented on lib/pure/unicode.nim:
```diff
@@ -157,6 +157,25 @@ proc `$`*(runes: seq[Rune]): string =
+  while i < pos:
+    o += runeLenAt(s, o)
+    inc i
+  o
```
use 'result' instead

Araq commented on lib/pure/unicode.nim:
```diff
@@ -157,6 +157,25 @@ proc `$`*(runes: seq[Rune]): string =
+    inc i
+  o
+
+proc rune*(s: string, pos:int): Rune =
```
pick a better name

oderwat commented:
I changed some names and hopefully made the behavior more consistent. I also added a PHP style "substr()" proc which allows this to be "fairly optimized". **I am still working on an better optimized  backward offset part which does length and offset on one run.**

  let s = "Hänsel  ««: 10,00€"

  echo "'", runeStrAtPos(s, 0), "'" # 'H'
  echo "'", runeStrAtPos(s, 10), "'" # ':'
  echo "'", runeStrAtPos(s, 9), "'" # '«'
  echo "'", runeStrAtPos(s, 17), "'" # '€' 
  #echo "'", runeStrAtPos(s, 18), "'" # index error
  #echo "'", runeStrAtPos(s, -1), "'" # does not compile (or range exception)

  echo "'", runeSubStr(s, 0), "'" # 'Hänsel  ««: 10,00€'
  echo "'", runeSubStr(s, 10), "'" # ': 10,00€'
  echo "'", runeSubStr(s, 18), "'" # ''
  echo "'", runeSubStr(s, 0, 10), "'" # 'Hänsel  ««'

  echo "'", runeSubStr(s, -6), "'" # '10,00€'
  echo "'", runeSubStr(s, 12), "'" # '10,00€'
  echo "'", runeSubStr(s, -6, 5), "'" # '10,00'
  echo "'", runeSubStr(s, -6, -1), "'" # '10,00'
  echo "'", runeSubStr(s, 12, -1), "'" # '10,00'
  echo "'", runeSubStr(s, 12, 5), "'" # '10,00'

  echo "'", runeSubStr(s, -18), "'" # 'Hänsel  ««: 10,00€'
  echo "'", runeSubStr(s, 0, 100), "'" # 'Hänsel  ««: 10,00€'
  echo "'", runeSubStr(s, 0, -100), "'" # ''
  echo "'", runeSubStr(s, -100, 117), "'" # 'Hänsel  ««: 10,00€'
```

oderwat commented:
I added an special proc which helps with optimizing "backward" offsets into an UTF-8 string. It does not really goes backward (I think that is possible I don't like the idea in general).

Probably runeStrAtPos() could left out as it is "nearly" the same as runeSubStr(s, x, 1) and the very seldom use case "may" not justify a specialized version. But then... it is already there.

Generally I think that the code has use cases and that rune offsetting is useful when working with UTF-8 strings of ISO-88.. origin. Which is what a lot of UTF-8 is. Like latin1 > UTF-8 as I show here in the examples and use in the tests.

oderwat commented on lib/pure/unicode.nim:
```diff
@@ -157,6 +157,107 @@ proc `$`*(runes: seq[Rune]): string =
+      result = s[o.. s.len-1]
+    elif len < 0:
+      let (e, rl) = runeReverseOffset(s, -len)
+      discard rl
```
I don't like this and could go away if somebody (or me) finds time for partial tuple unpacking :)

dom96 replied:
I think you can change this now to a ``_``.

Araq commented:
Every new proc here needs usage examples in its documentation and even then I begin to wonder if that shouldn'be in an unicode_quirky module instead. ;-)

Varriount commented:
@oderwat Examples?

Araq commented:
Please reopen if it's ready.
